monitoring/timedrift_test.go

```diff
@@ -0,0 +1,189 @@
+				s.c.successProbe(node2, 100*time.Millisecond),
+				s.c.successProbe(node3, 200*time.Millisecond),
+			},
+		},
+		{
+			Comment: "Time drift to both nodes is acceptable, one node is lagging behind",
+			Nodes:   []serf.Member{node1, node2, node3},
+			Times: map[string]time.Time{
+				node2.Name: s.clock.Now().Add(100 * time.Millisecond),
+				node3.Name: s.clock.Now().Add(-100 * time.Millisecond),
+			},
+			// Since we're using frozen time, latency will be 0 so
+			// time drift will be exactly the amount we specified.
+			Result: []*pb.Probe{
+				s.c.successProbe(node2, 100*time.Millisecond),
+				s.c.successProbe(node3, -100*time.Millisecond),
+			},
+		},
+		{
+			Comment: "Time drift to node-3 is high",
+			Nodes:   []serf.Member{node1, node2, node3},
+			Times: map[string]time.Time{
+				node2.Name: s.clock.Now().Add(100 * time.Millisecond),
+				node3.Name: s.clock.Now().Add(500 * time.Millisecond),
+			},
+			// Since we're using frozen time, latency will be 0 so
+			// time drift will be exactly the amount we specified.
+			Result: []*pb.Probe{
+				s.c.successProbe(node2, 100*time.Millisecond),
+				s.c.failureProbe(node3, 500*time.Millisecond),
+			},
+		},
+		{
+			Comment: "Time drift to node-2 is high",
+			Nodes:   []serf.Member{node1, node2, node3},
+			Times: map[string]time.Time{
+				node2.Name: s.clock.Now().Add(-time.Second),
+				node3.Name: s.clock.Now().Add(100 * time.Millisecond),
+			},
+			// Since we're using frozen time, latency will be 0 so
+			// time drift will be exactly the amount we specified.
+			Result: []*pb.Probe{
+				s.c.failureProbe(node2, -time.Second),
+				s.c.successProbe(node3, 100*time.Millisecond),
+			},
+		},
+		{
+			Comment: "Time drift to both nodes is high",
+			Nodes:   []serf.Member{node1, node2, node3},
+			Times: map[string]time.Time{
+				node2.Name: s.clock.Now().Add(-time.Second),
+				node3.Name: s.clock.Now().Add(time.Second),
+			},
+			// Since we're using frozen time, latency will be 0 so
+			// time drift will be exactly the amount we specified.
+			Result: []*pb.Probe{
+				s.c.failureProbe(node2, -time.Second),
+				s.c.failureProbe(node3, time.Second),
+			},
+		},
+	}
+	for _, test := range tests {
+		clients, err := ttlmap.New(10)
+		c.Assert(err, check.IsNil)
+		for nodeName, nodeTime := range test.Times {
+			clients.Set(
+				nodes[nodeName].Addr.String(),
+				newMockedTimeAgentClient(nodeTime),
+				time.Hour)
+		}
+		checker := &timeDriftChecker{
+			TimeDriftCheckerConfig: TimeDriftCheckerConfig{
+				Clock: s.clock,
+			},
+			FieldLogger: logrus.WithField(trace.Component, "test"),
+			clients:     clients,
+			serfClient:  agent.NewMockSerfClient(test.Nodes, nil),
+			name:        node1.Name,
+			self:        node1,
+		}
+		var probes health.Probes
+		checker.Check(context.TODO(), &probes)
+		c.Assert(probes.GetProbes(), check.DeepEquals, test.Result,
+			check.Commentf(test.Comment))
+	}
+}
+
+type mockedTimeAgentClient struct {
+	agent.Client
```

Comment:
What's this for? Ah, I get it - you need the full spectrum of the client APIs. Well, in this case I'd suggest to ask only for what's needed - i.e. if you only need the timing aspect from the agent, then why not have just this instead of the full-blown client:
```go
type Timer interface {
	Time(context.Context, *pb.TimeRequest) error
}
```
and then there will be no need for any other APIs.

Reply:
TBH I never really understood/liked this approach of making a gazillion of single-method interfaces just to limit the contract for something. We're not writing a standard library, and I think in our case it only increases the complexity / ability to explore the code.

Reply:
Ok, I also think a bogus field to satisfy a bigger interface is a superior solution.

Reply:
Sarcasm detected 😄I understand this is not ideal. I think I'm gonna remove this field and just provide stub implementations for the rest of methods.